tests/test_pytypes.py

```diff
@@ -541,3 +541,45 @@ def test_pass_bytes_or_unicode_to_string_types():
     else:
         with pytest.raises(TypeError):
             m.pass_to_pybind11_str(malformed_utf8)
+
+
+def test_weakref():
+    from weakref import getweakrefcount
+
+    # Apparently, you cannot weakly reference an object()
+    class WeaklyReferenced(object):
+        pass
+
+    def callback(wr):
+        # No `nonlocal` in Python 2
+        callback.called += 1
+
+    callback.called = 0
```

Comment:
Resetting `callback.called` only once makes the tests below order dependent. If someone has a reason to shuffle the test, they may get surprised. I'd move this down do just above the the `weakref_from_*_and_function` calls.

But I think I'd go one step further, using `@pytest.mark.parametrize` to reduce copy-pasting, that would automatically avoid the order dependence.

Reply:
Hmm, maybe. I tried something. I hope this is still readable enough to track down when one of the two tests fails, but it's shorter, at least.